pgmax/groups/enumeration.py

```diff
@@ -0,0 +1,459 @@
+        factor_configs: Array of shape (num_val_configs, num_variables)
+            An array containing explicit enumeration of all valid configurations
+    """
+
+    factor_configs: np.ndarray = field(init=True)
+    log_potentials: np.ndarray = field(init=True, default=np.empty((0,)))
+    factor_type: Type = field(init=False, default=enumeration.EnumerationFactor)
+
+    def __post_init__(self):
+        super().__post_init__()
+
+        num_val_configs = self.factor_configs.shape[0]
+        if self.log_potentials.shape[0] == 0:
+            log_potentials = np.zeros((self.num_factors, num_val_configs), dtype=float)
+        else:
+            if self.log_potentials.shape != (
+                num_val_configs,
+            ) and self.log_potentials.shape != (
+                self.num_factors,
+                num_val_configs,
+            ):
+                raise ValueError(
+                    f"Expected log potentials shape: {(num_val_configs,)} or {(self.num_factors, num_val_configs)}. "
+                    f"Got {self.log_potentials.shape}."
+                )
+            log_potentials = np.broadcast_to(
+                self.log_potentials, (self.num_factors, num_val_configs)
+            )
+
+        object.__setattr__(self, "log_potentials", log_potentials)
+
+    def _get_variables_to_factors(
+        self,
+    ) -> OrderedDict[FrozenSet, enumeration.EnumerationFactor]:
+        """Function that generates a dictionary mapping set of connected variables to factors.
+        This function is only called on demand when the user requires it.
+
+        Returns:
+            A dictionary mapping all possible set of connected variables to different factors.
+        """
+        variables_to_factors = collections.OrderedDict(
+            [
+                (
+                    frozenset(self.variable_names_for_factors[ii]),
+                    enumeration.EnumerationFactor(
+                        variables=tuple(
+                            self.variable_group[self.variable_names_for_factors[ii]]
+                        ),
+                        configs=self.factor_configs,
+                        log_potentials=self.log_potentials[ii],
+                    ),
+                )
+                for ii in range(len(self.variable_names_for_factors))
+            ]
+        )
+        return variables_to_factors
+
+    def compile_wiring(
+        self, vars_to_starts: Mapping[nodes.Variable, int]
+    ) -> enumeration.EnumerationWiring:
+        """Compile EnumerationWiring for the EnumerationFactorGroup
+
+        Args:
+            vars_to_starts: A dictionary that maps variables to their global starting indices
+                For an n-state variable, a global start index of m means the global indices
+                of its n variable states are m, m + 1, ..., m + n - 1
+
+        Returns:
+             EnumerationWiring for the EnumerationFactorGroup
+        """
+        return _compile_factor_group_wiring(
```

Comment:
This compiling function is really the same as the `compile_wiring` function in individual enumeration factors. There should not be two different implementations of the same function. Since we are trying to consolidate, can we define a function in `factors.enumeration` that handles the wiring compilation for all 3 cases?

Reply:
In addition, it still feels wrong to have to repeat this function multiple times.

One thought for possibly defining this function only once (inside the parent class): we still take the compile then concatenate approach, but we create a functional implementation for wiring compilation for individual factors (which can be used without instantiating individual factors). We then wrap things in numba to do the individual compilation in parallel before the concatenation.

The same approach should also work for LogicalFactors (and in factor any new factors we add) without change.

We can discuss more and try to tackle this in the next PR. For this PR fine to just consolidate all 3 places into one function.

Reply:
I agree that this function should only be called once.
I am planning to move this to numba in the next PR, let's keep it as it is for now